File: problem-solving/solutions/leetcode/blind75/src/main/kotlin/xyz/grind/coding/easy/GroupAnagrams.java
package xyz.grind.coding.easy;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GroupAnagrams {

    public static List<List<String>> groupAnagramsBruteForce(String[] strs){
        if(strs.length == 0) return new ArrayList<>();
        List<List<String>> groupedAnagrams = new ArrayList<List<String>>();
        Map<String, List<String>> computedAnagrams = new HashMap<String, List<String>>();
        for(String str : strs) {
            char[] strCharArray = str.toCharArray(); // because strings in java are immutable
            Arrays.sort(strCharArray);
            String anagramKey = new String(strCharArray);
            if(!computedAnagrams.containsKey(anagramKey)) computedAnagrams.put(anagramKey, new ArrayList());
            computedAnagrams.get(anagramKey).add(str);
        }
        groupedAnagrams.addAll(computedAnagrams.values());
        return groupedAnagrams;
    }

    public static List<List<String>> groupAnagramsBruteForceTwo(String[] strs){
        if(strs.length == 0) return new ArrayList<>();
        Map<String, List<String>> computedAnagrams = new HashMap<>();
        for(String str : strs) {
            char[] strCharArray = str.toCharArray(); // because strings in java are immutable
            Arrays.sort(strCharArray);
            String anagramKey = new String(strCharArray);
            if(!computedAnagrams.containsKey(anagramKey)) computedAnagrams.put(anagramKey, new ArrayList());
            computedAnagrams.get(anagramKey).add(str);
        }
        return new ArrayList(computedAnagrams.values());
    }

    public static List<List<String>> groupAnagrams(String[] strs) {
        if(strs.length == 0 ) return new ArrayList<>();
        List<List<String>> groupedAnagrams = new ArrayList<List<String>>();
        Map<String, List<String>> computedAnagram = new HashMap<String, List<String>>();
        for(String arrayAnagramElement : strs) {
            char[] strCounterCharArray = new char[26];
            for(char character : arrayAnagramElement.toCharArray()){
                strCounterCharArray[character - 'a']++;
            }
            String keyAnagram = new String(strCounterCharArray);
            if(!computedAnagram.containsKey(keyAnagram)) computedAnagram.computeIfAbsent(keyAnagram, k -> new ArrayList<>());
            computedAnagram.get(keyAnagram).add(arrayAnagramElement);
            System.out.println(computedAnagram);
        }

        groupedAnagrams.addAll(computedAnagram.values());
        return groupedAnagrams;
    }


}

File: problem-solving/solutions/leetcode/blind75/src/main/kotlin/xyz/grind/coding/easy/GroupAnagrams.kt
package xyz.grind.coding.easy

import java.util.Arrays

object GroupAnagrams {

    fun groupAnagramsBruteForce(strs: Array<String>): List<List<String>> {
        if (strs.isEmpty()) return ArrayList()
        val groupedAnagrams: MutableList<List<String>> = ArrayList()
        val computedAnagrams: MutableMap<String, MutableList<String>> = HashMap()
        for (str in strs) {
            val strCharArray = str.toCharArray() // because strings in java are immutable
            Arrays.sort(strCharArray)
            val anagramKey = String(strCharArray)
            if (!computedAnagrams.containsKey(anagramKey)) computedAnagrams[anagramKey] = arrayListOf()
            computedAnagrams[anagramKey]?.add(str)
        }
        groupedAnagrams.addAll(computedAnagrams.values)
        return groupedAnagrams
    }

    fun groupAnagramsBruteForceTwo(strs: Array<String>): List<List<String>> {
        if (strs.isEmpty()) return ArrayList()
        val computedAnagrams: MutableMap<String, MutableList<String>> = HashMap()
        for (str in strs) {
            val strCharArray = str.toCharArray() // because strings in java are immutable
            Arrays.sort(strCharArray)
            val anagramKey = String(strCharArray)
            if (!computedAnagrams.containsKey(anagramKey)) computedAnagrams[anagramKey] = ArrayList()
            computedAnagrams[anagramKey]!!.add(str)
        }
        return ArrayList(computedAnagrams.values)
    }

    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        if (strs.isEmpty()) return ArrayList()
        val groupedAnagrams: MutableList<List<String>> = ArrayList()
        val computedAnagram: MutableMap<String, MutableList<String>> = HashMap()
        for (arrayAnagramElement in strs) {
            val strCounterCharArray = CharArray(26)
            for (character in arrayAnagramElement.toCharArray()) {
                strCounterCharArray[character.code - 'a'.code]++
            }
            val keyAnagram = String(strCounterCharArray)
            if (!computedAnagram.containsKey(keyAnagram)) computedAnagram.computeIfAbsent(keyAnagram) { ArrayList() }
            computedAnagram[keyAnagram]?.add(arrayAnagramElement)
            println(computedAnagram)
        }
        groupedAnagrams.addAll(computedAnagram.values)
        return groupedAnagrams
    }
}
